src/databricks/labs/ucx/progress/jobs.py

```diff
@@ -48,7 +50,36 @@ def _job_problems(self) -> dict[int, list[str]]:
             index[job_problem.job_id].append(failure)
         return index
 
+    @cached_property
+    def _direct_fs_accesses(self) -> dict[str, list[str]]:
+        index = collections.defaultdict(list)
+        for crawler in self._direct_fs_access_crawlers:
+            for direct_fs_access in crawler.snapshot():
+                # The workflow and task source lineage are added by the WorkflowLinter
+                if len(direct_fs_access.source_lineage) < 2:
+                    continue
+                if direct_fs_access.source_lineage[0].object_type != "WORKFLOW":
+                    continue
+                if direct_fs_access.source_lineage[1].object_type != "TASK":
+                    continue
+                job_id = direct_fs_access.source_lineage[0].object_id
+                task_key = direct_fs_access.source_lineage[1].object_id  # <job id>/<task key>
+                # Follow same failure message structure as the JobProblems above and DirectFsAccessPyLinter deprecation
+                code = "direct-filesystem-access"
+                message = f"The use of direct filesystem references is deprecated: {direct_fs_access.path}"
+                failure = f"{code}: {task_key} task: {direct_fs_access.source_id}: {message}"
+                index[job_id].append(failure)
+        return index
```

Comment:
This shouldn't be part of a cached property because there's no way to reset it (or signal a refresh) and the encoder itself is a singleton.

Instead, look at how we do this in `TableEncoder`:

 - Override `.append_inventory_snapshot()`, so it loads the enrichment data.
 - Implement a new per-record encoder, for example `._encode_table_as_historical(record, extra, things, needed, for, enrichment)`
 - The new per-record encoder invokes `super().encode_record_as_historical(record)` and then performs enrichment.

This achieves the goals of only loading the data once (per snapshot encode) while not inadvertently keeping (potentially stale) data cached.

Reply:
I saw that and it is better yes, but decided not to rewrite it here to separate feature from refactor

Reply:
https://github.com/databrickslabs/ucx/issues/3377